src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class Instrumentation
+    {
+        private static MethodSymbol _createPayload = null;
+        private static MethodSymbol _flushPayload = null;
+
+        internal static BoundBlock InjectInstrumentation(MethodSymbol method, BoundBlock methodBody, int methodOrdinal, TypeCompilationState compilationState, CSharpCompilation compilation, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider, out ImmutableArray<SourceSpan> dynamicAnalysisSpans)
+        {
+            if (methodBody != null && (object)method != GetCreatePayload(compilation) && (object)method != GetFlushPayload(compilation))
```

Comment:
The checks are not necessary in the real world, in which the helper class comes from a referenced assembly. They are needed only when the helper class definition is part of the current compilation. (In particular, if GetCreatePayload is instrumented it goes into an infinite recursion.) I'll write something in the source text...